br/pkg/lightning/backend/local/engine.go

```diff
@@ -1008,25 +1006,29 @@ func (w *Writer) appendRowsSorted(kvs []common.KvPair) error {
 			return errors.Trace(err)
 		}
 		w.writer = writer
-		w.writer.minKey = append([]byte{}, kvs[0].Key...)
 	}
 
-	totalKeyLen := 0
-	for i := 0; i < len(kvs); i++ {
-		totalKeyLen += w.engine.keyAdapter.EncodedLen(kvs[i].Key)
-	}
-	buf := make([]byte, totalKeyLen)
-	encodedKvs := make([]common.KvPair, len(kvs))
+	keyAdapter := w.engine.keyAdapter
+	totalKeySize := 0
 	for i := 0; i < len(kvs); i++ {
-		encodedKey := w.engine.keyAdapter.Encode(buf, kvs[i].Key, kvs[i].RowID, kvs[i].Offset)
-		buf = buf[len(encodedKey):]
-		encodedKvs[i] = common.KvPair{Key: encodedKey, Val: kvs[i].Val}
-		w.batchSize += int64(len(encodedKvs[i].Key) + len(encodedKvs[i].Val))
+		keySize := keyAdapter.EncodedLen(kvs[i].Key)
+		w.batchSize += int64(keySize + len(kvs[i].Val))
+		totalKeySize += keySize
+	}
+	w.batchCount += len(kvs)
+	// noopKeyAdapter doesn't really change the key,
+	// skipping the encoding to avoid unnecessary alloc and copy.
+	if _, ok := keyAdapter.(noopKeyAdapter); !ok {
+		buf := make([]byte, 0, totalKeySize)
```

Comment:
I think we can put this but in `Writer` for reusing to avoid alloc for each call.

Reply:
Done.